File: LeetCode/iter1/c++/HouseRobberIII.cpp
#include <iostream>

using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

int robHelper(TreeNode *root,int &left,int &right) 
{
	if (root == NULL)
		return 0;

	int leftleft = 0, leftright = 0, rightLeft = 0, rightright = 0;
	left = robHelper(root->left,leftleft,leftright);
	right = robHelper(root->right, rightLeft,rightright);

	if (left + right > root->val + leftleft + leftright + rightLeft + rightright)
		return left + right;
	else
		return root->val + leftleft + leftright + rightLeft + rightright;
}

int rob(TreeNode* root) 
{
	int left = 0, right = 0;
	int result = robHelper(root,left,right);

	return result;
}

int main()
{
	TreeNode *root = new TreeNode(2);
	TreeNode *node1 = new TreeNode(1);
	TreeNode *node2 = new TreeNode(3);
	TreeNode *node3 = new TreeNode(4);
	//TreeNode *node4 = new TreeNode(1);
	root->left = node1;
	root->right = node2;

	node1->right = node3;
	//node2->left = node4;

	cout << rob(root);


	return 0;
}

File: LeetCode/iter1/python/HouseRobberIII.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def robHelper(self,root):
        if(root == None):
            return [0,0]
        
        leftL = self.robHelper(root.left)
        rightL = self.robHelper(root.right)
        
        leftLmax = leftL[0]
        if(leftLmax < leftL[1]):
            leftLmax = leftL[1]
        
        rightLmax = rightL[0]
        if(rightLmax < rightL[1]):
            rightLmax = rightL[1]
        
        fatherVal = leftLmax + rightLmax 
        grandChildVal = root.val + leftL[1] + rightL[1]
        
        return [grandChildVal,fatherVal]
    
    def rob(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        result = []
        result = self.robHelper(root)
        if(result[0] > result[1]):
            return result[0]
        else:
            return result[1]
        
        
#测试程序
s = Solution()
root = TreeNode(3)
node1 = TreeNode(2)
node2 = TreeNode(3)
node3 = TreeNode(3)
node4 = TreeNode(1)

root.left = node1
root.right = node2
node1.right = node3
node2.right = node4

print(s.rob(root))